internal/certgen/certgen.go

```diff
@@ -0,0 +1,331 @@
+}
+
+type certData struct {
+	Filename string
+	Data     []byte
+}
+
+// NewContourCerts generates a new ContourCerts with default values.
+// NOTE(youngnick) This means that there's only one place all these names
+// are hard-coded.
+func NewContourCerts() *ContourCerts {
+	return &ContourCerts{
+		CA: &keyPair{
+			SecretName: "cacert",
+			Cert: &certData{
+				Filename: "CAcert.pem",
+			},
+			Key: &certData{
+				Filename: "CAkey.pem",
+			},
+		},
+		Contour: &keyPair{
+			SecretName: "contourcert",
+			Cert: &certData{
+				Filename: "contourcert.pem",
+			},
+			Key: &certData{
+				Filename: "contourkey.pem",
+			},
+		},
+		Envoy: &keyPair{
+			SecretName: "envoycert",
+			Cert: &certData{
+				Filename: "envoycert.pem",
+			},
+			Key: &certData{
+				Filename: "envoykey.pem",
+			},
+		},
+	}
+}
+
+func (kp *keyPair) writePEMs(outputDir string) error {
+	err := kp.Cert.writePEM(outputDir)
+	if err != nil {
+		return err
+	}
+	return kp.Key.writePEM(outputDir)
+}
+
+func (cd *certData) writePEM(outputDir string) error {
+	return dumpFile(outputDir+"/"+cd.Filename, cd.Data)
+
+}
+
+// writeCertPEMs writes out all the PEMs for all the certs, using
+// the stored filenames.
+// TODO(youngnick) we should be able to use a similar pattern for the
+// secrets, hopefully.
+func (cc *ContourCerts) writeCertPEMs(outputDir string) error {
+	err := cc.CA.writePEMs(outputDir)
+	if err != nil {
+		return err
+	}
+	err = cc.Contour.writePEMs(outputDir)
+	if err != nil {
+		return err
+	}
+	return cc.Envoy.writePEMs(outputDir)
+
+}
+
+// GenerateCerts performs the actual cert generation steps and then returns the certs for the output functions.
+func GenerateCerts(certConfig *Config) (*ContourCerts, error) {
+
+	now := time.Now()
+	expiry := now.Add(24 * 365 * time.Hour)
+	caKeyPEM, caCertPEM, err := newCA(rand.Reader, "Project Contour", expiry)
+	if err != nil {
+		return nil, err
+	}
+
+	contourKey, contourCert, err := newCert(rand.Reader,
+		caCertPEM,
+		caKeyPEM,
+		expiry,
+		"contour",
+		certConfig.Namespace,
+	)
+	if err != nil {
+		return nil, err
+	}
+	envoyKey, envoyCert, err := newCert(rand.Reader,
+		caCertPEM,
+		caKeyPEM,
+		expiry,
+		"envoy",
+		certConfig.Namespace,
+	)
+	if err != nil {
+		return nil, err
+	}
+	newCerts := NewContourCerts()
+	newCerts.CA.Cert.Data = caCertPEM
+	newCerts.CA.Key.Data = caKeyPEM
+	newCerts.Contour.Cert.Data = contourCert
+	newCerts.Contour.Key.Data = contourKey
+	newCerts.Envoy.Cert.Data = envoyCert
+	newCerts.Envoy.Key.Data = envoyKey
+
+	return newCerts, nil
+
+}
+
+// OutputCerts outputs the certs in certs as directed by config.
+func OutputCerts(certgenConfig *Config, certs *ContourCerts) error {
+	log := certgenConfig.FieldLogger
+
+	if certgenConfig.OutputPEM {
+		// TODO(youngnick): Should we sanitize this value?
+		err := os.MkdirAll(certgenConfig.OutputDir, 0755)
+		if err != nil {
+			return err
+		}
+
+		log.Infof("Outputting certs to PEM files in %s/", certgenConfig.OutputDir)
```

Comment:
I'd drop the logger here and just use straigth fmt.Printf for two reasons

1. log goes to stderr
2. nobody needs timestamps in their "here's your files" line